client/consul.go

```diff
@@ -0,0 +1,159 @@
+// Care must always be taken to do type checks when casting, as structure could
+// potentially change over time.
+type Self = map[string]map[string]interface{}
+
+// NewConsulClient constructs a consul api client
+func NewConsulClient(conf *config.Config, maxRetry int) (*ConsulClient, error) {
+	consulConf := conf.Consul
+	transport := hcat.TransportInput{
+		SSLEnabled: *consulConf.TLS.Enabled,
+		SSLVerify:  *consulConf.TLS.Verify,
+		SSLCert:    *consulConf.TLS.Cert,
+		SSLKey:     *consulConf.TLS.Key,
+		SSLCACert:  *consulConf.TLS.CACert,
+		SSLCAPath:  *consulConf.TLS.CAPath,
+		ServerName: *consulConf.TLS.ServerName,
+
+		DialKeepAlive:       *consulConf.Transport.DialKeepAlive,
+		DialTimeout:         *consulConf.Transport.DialTimeout,
+		DisableKeepAlives:   *consulConf.Transport.DisableKeepAlives,
+		IdleConnTimeout:     *consulConf.Transport.IdleConnTimeout,
+		MaxIdleConns:        *consulConf.Transport.MaxIdleConns,
+		MaxIdleConnsPerHost: *consulConf.Transport.MaxIdleConnsPerHost,
+		TLSHandshakeTimeout: *consulConf.Transport.TLSHandshakeTimeout,
+	}
+
+	consul := hcat.ConsulInput{
+		Address:      *consulConf.Address,
+		Token:        *consulConf.Token,
+		AuthEnabled:  *consulConf.Auth.Enabled,
+		AuthUsername: *consulConf.Auth.Username,
+		AuthPassword: *consulConf.Auth.Password,
+		Transport:    transport,
+	}
+
+	clients := hcat.NewClientSet()
+	if err := clients.AddConsul(consul); err != nil {
+		return nil, err
+	}
+
+	logger := logging.Global().Named(loggingSystemName).Named(consulSubsystemName)
+
+	r := retry.NewRetry(maxRetry, time.Now().UnixNano())
+	return &ConsulClient{
+		Client: clients.Consul(),
+		retry:  r,
+		logger: logger}, nil
+}
+
+// GetLicense queries Consul for a signed license, and returns it if available
+func (c *ConsulClient) GetLicense(ctx context.Context, q *consulapi.QueryOptions) (string, error) {
+	c.logger.Debug("getting license")
+
+	desc := "consul client get license"
+	var err error
+	var license string
+
+	err = c.retry.Do(ctx, func(context.Context) error {
+		license, err = c.Operator().LicenseGetSigned(q)
+		if err != nil {
+			license = ""
+			return err
+		}
+		return nil
+	}, desc)
+
+	return license, err
+}
+
+// GetSKU queries Consul for information about itself, it then
+// parses this information to determine whether the Consul being
+// queried is Enterprise or OSS
+func (c *ConsulClient) GetSKU(ctx context.Context) (string, error) {
+	c.logger.Debug("getting sku")
+
+	desc := "consul client get sku"
+	var err error
+	var info Self
+
+	err = c.retry.Do(ctx, func(context.Context) error {
+		info, err = c.Agent().Self()
+		if err != nil {
+			info = nil
+			return err
+		}
+		return nil
+	}, desc)
+	if err != nil {
+		return "", err
+	}
+
+	sku, ok := parseSKU(info)
+	if !ok {
+		return "", errors.New("unable to parse sku")
+	}
+	return sku, nil
+}
+
+func parseSKU(info Self) (string, bool) {
+	v, ok := info["Config"]["Version"].(string)
```

Comment:
I'm not sure if there's any sensitive data in SKU? I think it could be useful to log or return an error containing the actual vs. expected SKU format for debugging purposes

Reply:
I'll look into this

Reply:
So it looks like the response doesn't include sensitive information (returns hidden, for example DebugConfig.ACLInitialManagementToken)


Here's and example:
```JSON
{
    "Config": {
        "Datacenter": "dc1",
        "PrimaryDatacenter": "dc1",
        "NodeName": "mwilkerson-C02FX04LMD6R",
        "NodeID": "ab7ec898-1bce-a45c-94b3-7e1ab5d12bf3",
        "Revision": "e319d7ed",
        "Server": true,
        "Version": "1.11.3"
    },
    "DebugConfig": {
        "ACLEnableKeyListPolicy": false,
        "ACLInitialManagementToken": "hidden",
        "ACLResolverSettings": {
            "ACLDefaultPolicy": "allow",
            "ACLDownPolicy": "extend-cache",
            "ACLPolicyTTL": "30s",
            "ACLRoleTTL": "0s",
            "ACLTokenTTL": "30s",
            "ACLsEnabled": false,
            "Datacenter": "dc1",
            "EnterpriseMeta": {},
            "NodeName": "mwilkerson-C02FX04LMD6R"
        },
        "ACLTokenReplication": false,
        "ACLTokens": {
            "ACLAgentRecoveryToken": "hidden",
            "ACLAgentToken": "hidden",
            "ACLDefaultToken": "hidden",
            "ACLReplicationToken": "hidden",
            "DataDir": "",
            "EnablePersistence": false,
            "EnterpriseConfig": {}
        },
        "ACLsEnabled": false,
        "AEInterval": "1m0s",
        "AdvertiseAddrLAN": "127.0.0.1",
        "AdvertiseAddrWAN": "127.0.0.1",
        "AdvertiseReconnectTimeout": "0s",
        "AllowWriteHTTPFrom": [],
        "AutoConfig": {
            "Authorizer": {
                "AllowReuse": false,
                "AuthMethod": {
                    "ACLAuthMethodEnterpriseFields": {},
                    "Config": {
                        "BoundAudiences": null,
                        "BoundIssuer": "",
                        "ClaimMappings": null,
                        "ClockSkewLeeway": 0,
                        "ExpirationLeeway": 0,
                        "JWKSCACert": "",
                        "JWKSURL": "",
                        "JWTSupportedAlgs": null,
                        "JWTValidationPubKeys": null,
                        "ListClaimMappings": null,
                        "NotBeforeLeeway": 0,
                        "OIDCDiscoveryCACert": "",
                        "OIDCDiscoveryURL": ""
                    },
                    "Description": "",
                    "DisplayName": "",
                    "EnterpriseMeta": {},
                    "MaxTokenTTL": "0s",
                    "Name": "Auto Config Authorizer",
                    "RaftIndex": {
                        "CreateIndex": 0,
                        "ModifyIndex": 0
                    },
                    "TokenLocality": "",
                    "Type": "jwt"
                },
                "ClaimAssertions": [],
                "Enabled": false
            },
            "DNSSANs": [],
            "Enabled": false,
            "IPSANs": [],
            "IntroToken": "hidden",
            "IntroTokenFile": "",
            "ServerAddresses": []
        },
        "AutoEncryptAllowTLS": false,
        "AutoEncryptDNSSAN": [],
        "AutoEncryptIPSAN": [],
        "AutoEncryptTLS": false,
        "AutopilotCleanupDeadServers": true,
        "AutopilotDisableUpgradeMigration": false,
        "AutopilotLastContactThreshold": "200ms",
        "AutopilotMaxTrailingLogs": 250,
        "AutopilotMinQuorum": 0,
        "AutopilotRedundancyZoneTag": "",
        "AutopilotServerStabilizationTime": "10s",
        "AutopilotUpgradeVersionTag": "",
        "BindAddr": "127.0.0.1",
        "Bootstrap": false,
        "BootstrapExpect": 0,
        "CAFile": "",
        "CAPath": "",
        "Cache": {
            "EntryFetchMaxBurst": 2,
            "EntryFetchRate": 1.7976931348623157e+308,
            "Logger": null
        },
        "CertFile": "",
        "CheckDeregisterIntervalMin": "1m0s",
        "CheckOutputMaxSize": 4096,
        "CheckReapInterval": "30s",
        "CheckUpdateInterval": "5m0s",
        "Checks": [],
        "ClientAddrs": [
            "127.0.0.1"
        ],
        "ConfigEntryBootstrap": [],
        "ConnectCAConfig": {},
        "ConnectCAProvider": "",
        "ConnectEnabled": true,
        "ConnectMeshGatewayWANFederationEnabled": false,
        "ConnectSidecarMaxPort": 21255,
        "ConnectSidecarMinPort": 21000,
        "ConnectTestCALeafRootChangeSpread": "0s",
        "ConsulCoordinateUpdateBatchSize": 128,
        "ConsulCoordinateUpdateMaxBatches": 5,
        "ConsulCoordinateUpdatePeriod": "100ms",
        "ConsulRaftElectionTimeout": "52ms",
        "ConsulRaftHeartbeatTimeout": "35ms",
        "ConsulRaftLeaderLeaseTimeout": "20ms",
        "ConsulServerHealthInterval": "10ms",
        "DNSARecordLimit": 0,
        "DNSAddrs": [
            "tcp://127.0.0.1:8600",
            "udp://127.0.0.1:8600"
        ],
        "DNSAllowStale": true,
        "DNSAltDomain": "",
        "DNSCacheMaxAge": "0s",
        "DNSDisableCompression": false,
        "DNSDomain": "consul.",
        "DNSEnableTruncate": false,
        "DNSMaxStale": "87600h0m0s",
        "DNSNodeMetaTXT": true,
        "DNSNodeTTL": "0s",
        "DNSOnlyPassing": false,
        "DNSPort": 8600,
        "DNSRecursorStrategy": "sequential",
        "DNSRecursorTimeout": "2s",
        "DNSRecursors": [],
        "DNSSOA": {
            "Expire": 86400,
            "Minttl": 0,
            "Refresh": 3600,
            "Retry": 600
        },
        "DNSServiceTTL": {},
        "DNSUDPAnswerLimit": 3,
        "DNSUseCache": false,
        "DataDir": "",
        "Datacenter": "dc1",
        "DefaultQueryTime": "5m0s",
        "DevMode": true,
        "DisableAnonymousSignature": true,
        "DisableCoordinates": false,
        "DisableHTTPUnprintableCharFilter": false,
        "DisableHostNodeID": true,
        "DisableKeyringFile": true,
        "DisableRemoteExec": true,
        "DisableUpdateCheck": false,
        "DiscardCheckOutput": false,
        "DiscoveryMaxStale": "0s",
        "EnableAgentTLSForChecks": false,
        "EnableCentralServiceConfig": true,
        "EnableDebug": true,
        "EnableLocalScriptChecks": false,
        "EnableRemoteScriptChecks": false,
        "EncryptKey": "hidden",
        "EncryptVerifyIncoming": true,
        "EncryptVerifyOutgoing": true,
        "EnterpriseRuntimeConfig": {},
        "ExposeMaxPort": 21755,
        "ExposeMinPort": 21500,
        "GRPCAddrs": [
            "tcp://127.0.0.1:8502"
        ],
        "GRPCPort": 8502,
        "GossipLANGossipInterval": "100ms",
        "GossipLANGossipNodes": 3,
        "GossipLANProbeInterval": "100ms",
        "GossipLANProbeTimeout": "100ms",
        "GossipLANRetransmitMult": 4,
        "GossipLANSuspicionMult": 3,
        "GossipWANGossipInterval": "100ms",
        "GossipWANGossipNodes": 3,
        "GossipWANProbeInterval": "100ms",
        "GossipWANProbeTimeout": "100ms",
        "GossipWANRetransmitMult": 4,
        "GossipWANSuspicionMult": 3,
        "HTTPAddrs": [
            "tcp://127.0.0.1:8500"
        ],
        "HTTPBlockEndpoints": [],
        "HTTPMaxConnsPerClient": 200,
        "HTTPMaxHeaderBytes": 0,
        "HTTPPort": 8500,
        "HTTPResponseHeaders": {},
        "HTTPSAddrs": [],
        "HTTPSHandshakeTimeout": "5s",
        "HTTPSPort": -1,
        "HTTPUseCache": true,
        "KVMaxValueSize": 524288,
        "KeyFile": "hidden",
        "LeaveDrainTime": "5s",
        "LeaveOnTerm": false,
        "Logging": {
            "EnableSyslog": false,
            "LogFilePath": "",
            "LogJSON": false,
            "LogLevel": "DEBUG",
            "LogRotateBytes": 0,
            "LogRotateDuration": "0s",
            "LogRotateMaxFiles": 0,
            "Name": "",
            "SyslogFacility": "LOCAL0"
        },
        "MaxQueryTime": "10m0s",
        "NodeID": "ab7ec898-1bce-a45c-94b3-7e1ab5d12bf3",
        "NodeMeta": {},
        "NodeName": "mwilkerson-C02FX04LMD6R",
        "PidFile": "",
        "PrimaryDatacenter": "dc1",
        "PrimaryGateways": [],
        "PrimaryGatewaysInterval": "30s",
        "RPCAdvertiseAddr": "tcp://127.0.0.1:8300",
        "RPCBindAddr": "tcp://127.0.0.1:8300",
        "RPCConfig": {
            "EnableStreaming": true
        },
        "RPCHandshakeTimeout": "5s",
        "RPCHoldTimeout": "7s",
        "RPCMaxBurst": 1000,
        "RPCMaxConnsPerClient": 100,
        "RPCProtocol": 2,
        "RPCRateLimit": -1,
        "RaftBoltDBConfig": {
            "NoFreelistSync": false
        },
        "RaftProtocol": 3,
        "RaftSnapshotInterval": "30s",
        "RaftSnapshotThreshold": 16384,
        "RaftTrailingLogs": 10240,
        "ReadReplica": false,
        "ReconnectTimeoutLAN": "0s",
        "ReconnectTimeoutWAN": "0s",
        "RejoinAfterLeave": false,
        "RetryJoinIntervalLAN": "30s",
        "RetryJoinIntervalWAN": "30s",
        "RetryJoinLAN": [],
        "RetryJoinMaxAttemptsLAN": 0,
        "RetryJoinMaxAttemptsWAN": 0,
        "RetryJoinWAN": [],
        "Revision": "e319d7ed",
        "SegmentLimit": 64,
        "SegmentName": "",
        "SegmentNameLimit": 64,
        "Segments": [],
        "SerfAdvertiseAddrLAN": "tcp://127.0.0.1:8301",
        "SerfAdvertiseAddrWAN": "tcp://127.0.0.1:8302",
        "SerfAllowedCIDRsLAN": [],
        "SerfAllowedCIDRsWAN": [],
        "SerfBindAddrLAN": "tcp://127.0.0.1:8301",
        "SerfBindAddrWAN": "tcp://127.0.0.1:8302",
        "SerfPortLAN": 8301,
        "SerfPortWAN": 8302,
        "ServerMode": true,
        "ServerName": "",
        "ServerPort": 8300,
        "Services": [
            {
                "Address": "",
                "Check": {
                    "AliasNode": "",
                    "AliasService": "",
                    "Body": "",
                    "CheckID": "",
                    "DeregisterCriticalServiceAfter": "0s",
                    "DockerContainerID": "",
                    "FailuresBeforeCritical": 0,
                    "FailuresBeforeWarning": 0,
                    "GRPC": "",
                    "GRPCUseTLS": false,
                    "H2PING": "",
                    "H2PingUseTLS": false,
                    "HTTP": "",
                    "Header": {},
                    "Interval": "0s",
                    "Method": "",
                    "Name": "",
                    "Notes": "",
                    "OutputMaxSize": 0,
                    "ProxyGRPC": "",
                    "ProxyHTTP": "",
                    "ScriptArgs": [],
                    "Shell": "",
                    "Status": "",
                    "SuccessBeforePassing": 0,
                    "TCP": "",
                    "TLSServerName": "",
                    "TLSSkipVerify": false,
                    "TTL": "0s",
                    "Timeout": "0s"
                },
                "Checks": [],
                "Connect": null,
                "EnableTagOverride": false,
                "EnterpriseMeta": {},
                "ID": "",
                "Kind": "",
                "Meta": {},
                "Name": "web",
                "Port": 80,
                "Proxy": null,
                "SocketPath": "",
                "TaggedAddresses": {},
                "Tags": [
                    "example"
                ],
                "Token": "hidden",
                "Weights": {
                    "Passing": 1,
                    "Warning": 1
                }
            },
            {
                "Address": "",
                "Check": {
                    "AliasNode": "",
                    "AliasService": "",
                    "Body": "",
                    "CheckID": "",
                    "DeregisterCriticalServiceAfter": "0s",
                    "DockerContainerID": "",
                    "FailuresBeforeCritical": 0,
                    "FailuresBeforeWarning": 0,
                    "GRPC": "",
                    "GRPCUseTLS": false,
                    "H2PING": "",
                    "H2PingUseTLS": false,
                    "HTTP": "",
                    "Header": {},
                    "Interval": "0s",
                    "Method": "",
                    "Name": "",
                    "Notes": "",
                    "OutputMaxSize": 0,
                    "ProxyGRPC": "",
                    "ProxyHTTP": "",
                    "ScriptArgs": [],
                    "Shell": "",
                    "Status": "",
                    "SuccessBeforePassing": 0,
                    "TCP": "",
                    "TLSServerName": "",
                    "TLSSkipVerify": false,
                    "TTL": "0s",
                    "Timeout": "0s"
                },
                "Checks": [],
                "Connect": null,
                "EnableTagOverride": false,
                "EnterpriseMeta": {},
                "ID": "",
                "Kind": "",
                "Meta": {},
                "Name": "api",
                "Port": 9000,
                "Proxy": null,
                "SocketPath": "",
                "TaggedAddresses": {},
                "Tags": [
                    "test"
                ],
                "Token": "hidden",
                "Weights": {
                    "Passing": 1,
                    "Warning": 1
                }
            }
        ],
        "SessionTTLMin": "0s",
        "SkipLeaveOnInt": true,
        "StartJoinAddrsLAN": [],
        "StartJoinAddrsWAN": [],
        "SyncCoordinateIntervalMin": "15s",
        "SyncCoordinateRateTarget": 64,
        "TLSCipherSuites": [],
        "TLSMinVersion": "tls12",
        "TLSPreferServerCipherSuites": false,
        "TaggedAddresses": {
            "lan": "127.0.0.1",
            "lan_ipv4": "127.0.0.1",
            "wan": "127.0.0.1",
            "wan_ipv4": "127.0.0.1"
        },
        "Telemetry": {
            "AllowedPrefixes": [],
            "BlockedPrefixes": [],
            "CirconusAPIApp": "",
            "CirconusAPIToken": "hidden",
            "CirconusAPIURL": "",
            "CirconusBrokerID": "",
            "CirconusBrokerSelectTag": "",
            "CirconusCheckDisplayName": "",
            "CirconusCheckForceMetricActivation": "",
            "CirconusCheckID": "",
            "CirconusCheckInstanceID": "",
            "CirconusCheckSearchTag": "",
            "CirconusCheckTags": "",
            "CirconusSubmissionInterval": "",
            "CirconusSubmissionURL": "",
            "Disable": false,
            "DisableCompatOneNine": false,
            "DisableHostname": false,
            "DogstatsdAddr": "",
            "DogstatsdTags": [],
            "FilterDefault": true,
            "MetricsPrefix": "consul",
            "PrometheusOpts": {
                "CounterDefinitions": [],
                "Expiration": "0s",
                "GaugeDefinitions": [],
                "Name": "consul",
                "Registerer": null,
                "SummaryDefinitions": []
            },
            "StatsdAddr": "",
            "StatsiteAddr": ""
        },
        "TranslateWANAddrs": false,
        "TxnMaxReqLen": 524288,
        "UIConfig": {
            "ContentPath": "/ui/",
            "DashboardURLTemplates": {},
            "Dir": "",
            "Enabled": true,
            "MetricsProvider": "",
            "MetricsProviderFiles": [],
            "MetricsProviderOptionsJSON": "",
            "MetricsProxy": {
                "AddHeaders": [],
                "BaseURL": "",
                "PathAllowlist": []
            }
        },
        "UnixSocketGroup": "",
        "UnixSocketMode": "",
        "UnixSocketUser": "",
        "UseStreamingBackend": true,
        "VerifyIncoming": false,
        "VerifyIncomingHTTPS": false,
        "VerifyIncomingRPC": false,
        "VerifyOutgoing": false,
        "VerifyServerHostname": false,
        "Version": "1.11.3",
        "VersionPrerelease": "",
        "Watches": []
    },
    "Coord": {
        "Vec": [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0
        ],
        "Error": 1.5,
        "Adjustment": 0,
        "Height": 0.00001
    },
    "Member": {
        "Name": "mwilkerson-C02FX04LMD6R",
        "Addr": "127.0.0.1",
        "Port": 8301,
        "Tags": {
            "acls": "0",
            "build": "1.11.3:e319d7ed",
            "dc": "dc1",
            "ft_fs": "1",
            "ft_si": "1",
            "id": "ab7ec898-1bce-a45c-94b3-7e1ab5d12bf3",
            "port": "8300",
            "raft_vsn": "3",
            "role": "consul",
            "segment": "",
            "vsn": "2",
            "vsn_max": "3",
            "vsn_min": "2",
            "wan_join_port": "8302"
        },
        "Status": 1,
        "ProtocolMin": 1,
        "ProtocolMax": 5,
        "ProtocolCur": 2,
        "DelegateMin": 2,
        "DelegateMax": 5,
        "DelegateCur": 4
    },
    "Stats": {
        "agent": {
            "check_monitors": "0",
            "check_ttls": "0",
            "checks": "0",
            "services": "2"
        },
        "build": {
            "prerelease": "",
            "revision": "e319d7ed",
            "version": "1.11.3"
        },
        "consul": {
            "acl": "disabled",
            "bootstrap": "false",
            "known_datacenters": "1",
            "leader": "true",
            "leader_addr": "127.0.0.1:8300",
            "server": "true"
        },
        "raft": {
            "applied_index": "15670",
            "commit_index": "15670",
            "fsm_pending": "0",
            "last_contact": "0",
            "last_log_index": "15670",
            "last_log_term": "2",
            "last_snapshot_index": "0",
            "last_snapshot_term": "0",
            "latest_configuration": "[{Suffrage:Voter ID:ab7ec898-1bce-a45c-94b3-7e1ab5d12bf3 Address:127.0.0.1:8300}]",
            "latest_configuration_index": "0",
            "num_peers": "0",
            "protocol_version": "3",
            "protocol_version_max": "3",
            "protocol_version_min": "0",
            "snapshot_version_max": "1",
            "snapshot_version_min": "0",
            "state": "Leader",
            "term": "2"
        },
        "runtime": {
            "arch": "amd64",
            "cpu_count": "16",
            "goroutines": "108",
            "max_procs": "16",
            "os": "darwin",
            "version": "go1.17.5"
        },
        "serf_lan": {
            "coordinate_resets": "0",
            "encrypted": "false",
            "event_queue": "1",
            "event_time": "2",
            "failed": "0",
            "health_score": "0",
            "intent_queue": "0",
            "left": "0",
            "member_time": "1",
            "members": "1",
            "query_queue": "0",
            "query_time": "1"
        },
        "serf_wan": {
            "coordinate_resets": "0",
            "encrypted": "false",
            "event_queue": "0",
            "event_time": "1",
            "failed": "0",
            "health_score": "0",
            "intent_queue": "0",
            "left": "0",
            "member_time": "1",
            "members": "1",
            "query_queue": "0",
            "query_time": "1"
        }
    },
    "Meta": {
        "consul-network-segment": ""
    },
    "xDS": {
        "SupportedProxies": {
            "envoy": [
                "1.20.1",
                "1.19.1",
                "1.18.4",
                "1.17.4"
            ]
        },
        "Port": 8502
    }
}
```

Reply:
I think it would be okay to include the response if the keys can't be found?

Reply:
If the key can't be found, though, does a user gain anything from seeing this full response? Especially since it's dense with information.

Reply:
Maybe a debug log for our purposes?

Reply:
Thanks for looking into this. Wow, that payload is pretty large. Agree with @mkam - it'll be hard for the user to dig through the payload to find the information needed to look into this problem. If there is actually an issue with parsing out the version though, the user will probably need to go query and examine the response 🤔 . I'm 50/50 and ok with skipping logging the response object

Reply:
I think a debug log would be okay, it would be useful for the case where it's failing and the user would be interested in performing the diagnostics/providing it to us a in a bug.